File: app/routes/services/llm/deepseekService.ts
import { HfInference } from "@huggingface/inference";

const isGreeting = (message: string): boolean => {
  const greetings = ["hi", "hello", "hey", "how are you", "good morning", "good evening", "good afternoon"];
  const cleanMessage = message.toLowerCase().trim();
  return greetings.some(greeting => 
    cleanMessage === greeting || 
    cleanMessage.startsWith(greeting + " ") || 
    cleanMessage.endsWith(" " + greeting)
  );
};

export const generateLLMResponse = async (prompt: string, products: any[] = []): Promise<{
  response: string; products?: any[] }> => {
  const hf = new HfInference(process.env.HUGGINGFACE_API_KEY);

  const userMessageMatch = prompt.match(/Question:\s*(.*?)(?:\n|$)/);
  const userMessage = userMessageMatch ? userMessageMatch[1] : prompt;

  if (isGreeting(userMessage)) {
    return { 
      response: `How can I assist you today? If you have any questions about our products, feel free to ask! I'm here to help you with all things!`,
      products: [] 
    };
  }  

  try{
    const response = await hf.textGeneration({
      model: "deepseek-ai/DeepSeek-R1-Distill-Qwen-32B",
      inputs: prompt,
      parameters: {
        max_new_tokens: 200,
        temperature: 0.3,
        return_full_text: false,
      },
    });

    let cleanedResponse = response.generated_text
      .replace(/<\/?think>|<\/?reasoning>/gi, '') 
      .replace(/^\s*\n/gm, '') 
      .trim();

    if (cleanedResponse.includes('</think>') || cleanedResponse.length > 300) {
      const lastParagraph = cleanedResponse.split('\n\n').pop() || cleanedResponse;
      cleanedResponse = lastParagraph.trim();
    }

    const lowerUserMessage = userMessage.toLowerCase();
    
    const isProductQuery = lowerUserMessage.includes('product') || 
                          lowerUserMessage.includes('show me') || 
                          lowerUserMessage.includes('list') || 
                          lowerUserMessage.includes('items') || 
                          products.some(p => lowerUserMessage.includes(p.product.toLowerCase().split('no description')[0].trim()));
    
    if (isProductQuery) {
      const transformedProducts = products
        .filter(product => {
          if (lowerUserMessage.includes('show') || lowerUserMessage.includes('list')) {
            return true; 
          } else {
            const productName = product.product.toLowerCase().split('no description')[0].trim();
            return lowerUserMessage.includes(productName);
          }
        })
        .map(product => {
          const productParts = product.product.split('No description available');
          const title = productParts[0].trim();
          const price = productParts[1] ? `$${productParts[1].trim()}` : 'Price not available';
          
          return {
            id: product.id,
            title: title,
            price: price,
            url: product.url,
            image: product.image
          };
      });

      return { response: cleanedResponse, products: transformedProducts };
    } else {
      return { response: cleanedResponse, products: [] }
    }
  } catch(error) {
    console.error("Error generating response:", error);
    return { response: "I don't have much information on this." };
  }
};

export const createDeepseekPrompt = (userMessage: string, contextTexts?: string): string => {
  return `
     You are a specialized product assistant that helps users find products from a catlog.

     Instructions:
     1. If the user asks for products (e.g., "Show me snowboards" or "List products"), provide a list of products from the catalog.
     2. If the user asks about a specific product (e.g., "Tell me about the Green Snowboard" or "Snowboard"), provide details about that product.
     3. If the user asks about features or specifications of a product, provide only factual information from the catalog.
     4. If no products match the user's query, respond with: "I couldn't find any products matching your request..."
     5. Do not make assumptions about products not in the catalog.
     6. Keep responses concise and focused on the products.
     7. Do not use phrases like "based on the Product Catlog" or "according to the information provided."
     8. Do not include thinking tags or explanations of your reasoning process.

    Product Catlog: ${contextTexts}

    User: ${userMessage}
  `;
};

File: api/services/llm/deepseek_service.py
import os
import re
import sys
from typing import List, Dict, Optional
from huggingface_hub import InferenceClient
from schemas.models import LLMResponse
from services.llm.langfuse_observation_module import langfuse_tracker
sys.path.append(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))

async def generate_llm_response(prompt: str, products: List[Dict] = []) -> LLMResponse:
    hf = InferenceClient(token=os.getenv("HUGGINGFACE_API_KEY"))

    user_message_match = re.search(r"Question:\s*(.*?)(?:\n|$)", prompt)
    user_message = user_message_match.group(1) if user_message_match else prompt

    try:
        response = hf.text_generation(
            model="deepseek-ai/DeepSeek-R1-Distill-Qwen-32B",
            prompt=prompt, 
            max_new_tokens=200,
            temperature=0.3,
            do_sample=True,
            return_full_text=False,
        )
       
        cleaned_response = response
        cleaned_response = re.sub(r'<\/?think>|<\/?reasoning>', '', cleaned_response, flags=re.IGNORECASE)
        cleaned_response = re.sub(r'^\s*\n', '', cleaned_response, flags=re.MULTILINE).strip()

        if '</think>' in cleaned_response or len(cleaned_response) > 300:
            last_paragraph = cleaned_response.split('\n\n')[-1] if '\n\n' in cleaned_response else cleaned_response
            cleaned_response = last_paragraph.strip()

        token_efficiency = langfuse_tracker.calculate_token_efficiency(cleaned_response)

        langfuse_tracker.track_llm_response(
            prompt=prompt, 
            response=cleaned_response, 
            metadata={
                "user_message": user_message,
                "token_efficiency": token_efficiency
            }
        )

        lower_user_message = user_message.lower()
        
        is_product_query = (
            'product' in lower_user_message or 
            'show me' in lower_user_message or 
            'list' in lower_user_message or 
            'items' in lower_user_message or 
            any(p['product'].lower().split('no description')[0].strip() in lower_user_message for p in products)
        )
        
        if is_product_query:
            transformed_products = []
            for product in products:
                if 'show' in lower_user_message or 'list' in lower_user_message:
                    include = True
                else:
                    product_name = product['title'].lower().strip()
                    include = product_name in lower_user_message
                
                if include:
                    transformed_products.append({
                        "id": product["id"],
                        "title": product["title"],
                        "price": product["price"],
                        "url": product["url"],
                        "image": product["image"]
                    })

            return LLMResponse(response=cleaned_response, products=transformed_products)
        else:
            return LLMResponse(response=cleaned_response, products=[])
    except Exception as error:
        print(f"Error generating response: {error}")
        return LLMResponse(response="We are currently not avilable, try again later!")


def create_deepseek_prompt(user_message: str, context_texts: Optional[str] = None) -> str:
    return f"""
        You are an expert Shopify assistant. Based on the user query and the retrieved store data, provide a    concise, accurate, and structured response. If the query is unclear or data is missing, say: [Unable to answer due to insufficient information]. Redirect ALL conversations to product queries.

        Your responses must be:
        - Concise (1-2 sentences max for greetings, 3-5 bullet points for products)
        - Strictly product-focused
        - NEVER explain your thought process
        - Free of repetition
        - Zero internal reasoning
        - No self-references ("I", "we")
        - Never add symbols excluding price

        1. For greetings/generic queries:
           If the user greets or asks generic questions, respond by:
            a) Acknowledging briefly
            b) Prompting to ask about products
            c) Providing 2-3 example product categories from the catalog

        2. For product queries
           a) ONLY use details from: {context_texts or 'NO CATALOG PROVIDED'}
           b) Format responses as:
              "Here are [N] matching products:"
                [Product] - [Price]
                [Product] - [Price]
              "Let me know if you'd like details on any!"
           b) If no match: "I couldn't find matches. Try these categories: [Category1], [Category2]"

        3. Never:
           - Repeat categories/products
           - Add unsolicited greetings to product responses
           - Exceed 5 bullet points
           - Never invent features/prices. Never say "based on the catalog."

        4. Data Source Requirements:
            a) ALL product responses MUST use EXACTLY these fields from {context_texts}:
                 - Product Name (exact match)
                 - Price (format: ₹XXX.XX)
            b) If {context_texts} is empty, respond ONLY with:
               "No product catalog available"

        5. Out of Context
            - Don't repeat or show the products
            - Your response should be:
            I aplogized, I don't have information on this...

        Good Example 1 (Greeting):
        User: Hi there!
        Assistant: Hello! How can I help you today?

        Good Example 2 (Generic Question):
        User: Show some products
        Assistant: We have a wide range of products are you looking for [Category 1] [Category 2] [Category 3]

        Good Product Query:
        User: Show snowboards
        Assistant: 
        Here are two snowboards:
          Snowboard Pro X - $299.99
          All-Mountain Board - $249.99
        Let me know if you'd like details on any!

        Good Example 2 (Generic Question):
        User: How are you?
        Assistant: I'm here to help with product questions!

        Good Example 3 (Product Query):
        User: Show me DSLR cameras
        Assistant: We have: 1) Nikon D850 ($2,999), 2) Canon EOS R5 ($3,899)

        Bad Product Query (Violates Rules):
        User: Show snowboards
        Assistant: Hello! Here are snowboards... [Greeting repeated]

        Bad Example (Generic Response):
        User: What's the weather today?
        Assistant: Sorry, I only handle product queries. Ask about: 
          Smartwatches
          Fitness trackers 

        Invalid Example (Leaks Thinking):
        User: Show jackets
        Response: Thinking... I found 2 jackets: [REDACTED]

        User: {user_message}
        Assistant:
    """